### PR TITLE
clusterctl - Add phase for get-kubeconfig

### PR DESCRIPTION
**Pull request by detiber**
**What this PR does / why we need it**:

Adds a phase for retrieving the kubeconfig for the target cluster

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #554 

**Release note**:
```release-note
clusterctl now has an 'alpha phases get-kubeconfig' command for retrieving the kubeconfig for a target cluster
```

### CODE REVIEW

File: cmd/clusterctl/cmd/alpha_phase_get_kubeconfig.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+			exitWithHelp(cmd, "Please specify a cluster name.")
+		}
+
+		if err := RunAlphaPhaseGetKubeconfig(pgko); err != nil {
+			klog.Exit(err)
+		}
+	},
+}
+
+func RunAlphaPhaseGetKubeconfig(pgko *AlphaPhaseGetKubeconfigOptions) error {
```
Comment (neolit123) [inline cmd/clusterctl/cmd/alpha_phase_get_kubeconfig.go]: would need an export comment.

File: cmd/clusterctl/phases/getkubeconfig.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/util"
+)
+
+const (
+	retryKubeConfigReady   = 10 * time.Second
+	timeoutKubeconfigReady = 20 * time.Minute
+)
+
+func GetKubeconfig(bootstrapClient clusterclient.Client, provider provider.ProviderDeployer, kubeconfigOutput string, clusterName, namespace string) (string, error) {
```
Comment (neolit123) [inline cmd/clusterctl/phases/getkubeconfig.go]: would need an export comment.

File: cmd/clusterctl/phases/getkubeconfig.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		}
+		kubeconfig = k
+		return true, nil
+	})
+
+	return kubeconfig, err
+}
+
+func writeKubeconfig(kubeconfig string, kubeconfigOutput string) error {
+	const fileMode = 0666
```
Comment (neolit123) [inline cmd/clusterctl/phases/getkubeconfig.go]: is r/w for all OK here?
`0660` is also an option.

File: cmd/clusterctl/phases/getkubeconfig.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		return "", err
+	}
+
+	klog.V(1).Info("Getting target cluster kubeconfig.")
+	targetKubeconfig, err := waitForKubeconfigReady(provider, cluster, master)
+	if err != nil {
+		return "", fmt.Errorf("unable to get target cluster kubeconfig: %v", err)
+	}
+
+	if err = writeKubeconfig(targetKubeconfig, kubeconfigOutput); err != nil {
```
Comment (vincepri) [inline cmd/clusterctl/phases/getkubeconfig.go]: This could be `:=`

File: cmd/clusterctl/clusterdeployer/clusterdeployer.go [hunk 5]
```diff
@@ -18,38 +18,16 @@ package clusterdeployer
...
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/clusterdeployer/clusterclient"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/clusterdeployer/provider"
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/phases"
 	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
-	"sigs.k8s.io/cluster-api/pkg/util"
 )
 
-// Deprecated interface for Provider specific logic. Please do not extend or add. This interface should be removed
-// once issues/158 and issues/160 below are fixed.
-type ProviderDeployer interface {
```
Comment (vincepri) [inline cmd/clusterctl/clusterdeployer/clusterdeployer.go]: Some providers test the interface in this package, the PR will need a release note to make sure folks update the interface package.

File: cmd/clusterctl/clusterdeployer/provider/provider.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+package provider
+
+import (
+	"k8s.io/client-go/kubernetes"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+// Deprecated interface for Provider specific logic. Please do not extend or add. This interface should be removed
+// once issues/158 and issues/160 below are fixed.
+type ProviderDeployer interface {
```
Comment (vincepri) [inline cmd/clusterctl/clusterdeployer/provider/provider.go]: Now that the package is named provider, I'd suggest naming this `Deployer`, as `provider.ProviderDeployer` stutters.

File: cmd/clusterctl/clusterdeployer/provider/provider.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	// TODO: This requirement can be removed once after: https://github.com/kubernetes-sigs/cluster-api/issues/158
+	GetIP(cluster *clusterv1.Cluster, machine *clusterv1.Machine) (string, error)
+	// TODO: This requirement can be removed after: https://github.com/kubernetes-sigs/cluster-api/issues/160
+	GetKubeConfig(cluster *clusterv1.Cluster, master *clusterv1.Machine) (string, error)
+}
+
+type ProviderComponentsStore interface {
```
Comment (vincepri) [inline cmd/clusterctl/clusterdeployer/provider/provider.go]: Same as above.

File: cmd/clusterctl/clusterdeployer/provider/provider.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	Save(providerComponents string) error
+	Load() (string, error)
+}
+
+type ProviderComponentsStoreFactory interface {
```
Comment (vincepri) [inline cmd/clusterctl/clusterdeployer/provider/provider.go]: Same as above.

**Comment by detiber** [comment]
Just finished testing with the latest changes (testing requires vendoring these changes into a provider implementation) and I was able to successfully pull a kubeconfig from an existing cluster.

File: cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go [hunk 9]
```diff
@@ -609,3 +609,39 @@ func ifErrRemove(pErr *error, path string) {
 		}
 	}
 }
+
+func GetClusterAPIObject(client Client, clusterName, namespace string) (*clusterv1.Cluster, *clusterv1.Machine, []*clusterv1.Machine, error) {
```
Comment (neolit123) [inline cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go]: minor nit: needs export comment.
do we have a linter in place that catches these?

**Comment by vincepri** [comment]
@neolit123 golint is enabled but comments are excluded. If it gets enabled, it'd require a lot of changed, gonna open an issue, wdyt?

**Comment by neolit123** [comment]
i'd say it's better to fix them all at once and then enable the linter on CI.
also these comments will be exported to godocs once cluster-api makes a more official release.

> gonna open an issue, wdyt?

SGTM

**Pull request merged**